File: frontend/src/features/work-record/model/workTypeDefinition.ts
import type { WorkType } from "@/entities/farm/types";

type WorkWorkflowKind =
  | "GENERIC"
  | "MOVEMENT"
  | "DISCARD"
  | "POTTING"
  | "STRUCTURE_CHANGE";

type WorkExecutionKind =
  | "MOVEMENT"
  | "DISCARD"
  | "POTTING"
  | "STRUCTURE_CHANGE";

export type WorkRecordResultKind = "DISCARD" | "POTTING" | "STRUCTURE_CHANGE";

type WorkTypeDefinition = {
  category: "GENERAL" | "STRUCTURE_CHANGE" | null;
  execution: WorkExecutionKind | null;
  planGuidance: string;
  planSupported: boolean;
  recordSupported: boolean;
  recordResult: WorkRecordResultKind | null;
  targetSource: "ORCHID_GROUP" | "INBOUND_RECORD";
  workflow: WorkWorkflowKind;
};

const GENERIC_DEFINITION: WorkTypeDefinition = {
  category: "GENERAL",
  execution: null,
  planGuidance:
    "난 묶음을 계획 대상으로 확정하고, 작업 유형에 맞는 기록 내용을 저장합니다.",
  planSupported: true,
  recordSupported: true,
  recordResult: null,
  targetSource: "ORCHID_GROUP",
  workflow: "GENERIC",
};

const SPECIAL_DEFINITIONS: Record<string, WorkTypeDefinition> = {
  MOVEMENT: {
    category: null,
    execution: "MOVEMENT",
    planGuidance:
      "원본 난 묶음을 계획 대상으로 확정하고, 실행할 때 각 묶음의 목적 구역과 위치를 입력합니다.",
    planSupported: true,
    recordSupported: false,
    recordResult: null,
    targetSource: "ORCHID_GROUP",
    workflow: "MOVEMENT",
  },
  DISCARD: {
    category: "STRUCTURE_CHANGE",
    execution: "DISCARD",
    planGuidance:
      "폐기할 난 묶음을 대상으로 정하고, 실행할 때 대상별 일부 또는 전량 폐기 수량과 사유를 입력합니다.",
    planSupported: true,
    recordSupported: true,
    recordResult: "DISCARD",
    targetSource: "ORCHID_GROUP",
    workflow: "DISCARD",
  },
  POTTING: {
    category: "STRUCTURE_CHANGE",
    execution: "POTTING",
    planGuidance:
      "포트 작업 대기 입고 기록을 선택하고, 실행할 때 실제 수량과 배치 위치를 입력합니다.",
    planSupported: true,
    recordSupported: true,
    recordResult: "POTTING",
    targetSource: "INBOUND_RECORD",
    workflow: "POTTING",
  },
  REPOT: structureChangeDefinition(
    "보통 자동·사용자 그룹 하나를 대상으로 정하고, 실행 회차마다 작업한 일부 수량과 여러 결과 묶음을 기록합니다.",
  ),
  DIVIDE: structureChangeDefinition(
    "대상 그룹을 정한 뒤 실행 회차마다 작업한 일부 수량과 여러 결과 묶음을 기록합니다.",
  ),
  MERGE: structureChangeDefinition(
    "같은 품종의 대상 그룹을 정하고, 실행 회차마다 작업한 일부 수량과 여러 결과 묶음을 기록합니다.",
  ),
};

export function getWorkTypeDefinition(
  workType: Pick<WorkType, "code"> | null | undefined,
): WorkTypeDefinition {
  return workType
    ? (SPECIAL_DEFINITIONS[workType.code] ?? GENERIC_DEFINITION)
    : GENERIC_DEFINITION;
}

export function getWorkExecutionKind(code: string): WorkExecutionKind | null {
  return SPECIAL_DEFINITIONS[code]?.execution ?? null;
}

function structureChangeDefinition(planGuidance: string): WorkTypeDefinition {
  return {
    category: "STRUCTURE_CHANGE",
    execution: "STRUCTURE_CHANGE",
    planGuidance,
    planSupported: true,
    recordSupported: true,
    recordResult: "STRUCTURE_CHANGE",
    targetSource: "ORCHID_GROUP",
    workflow: "STRUCTURE_CHANGE",
  };
}

File: backend/src/main/java/com/greenhouse/backend/work/domain/operation/WorkTypeDefinition.java
package com.greenhouse.backend.work.domain.operation;

import com.greenhouse.backend.work.domain.target.WorkTargetReferenceType;
import java.util.Arrays;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import lombok.RequiredArgsConstructor;

/** Stable work behavior; active/system flags and lifecycle state remain on the entities. */
@RequiredArgsConstructor
public enum WorkTypeDefinition {
	GENERIC(null, WorkTypeWorkflow.GENERIC, WorkTargetReferenceType.ORCHID_GROUP, true),
	INBOUND(null, WorkTypeWorkflow.GENERIC, WorkTargetReferenceType.ORCHID_GROUP, false),
	POTTING("POTTING", WorkTypeWorkflow.POTTING, WorkTargetReferenceType.INBOUND_RECORD, false),
	MOVEMENT(null, WorkTypeWorkflow.MOVEMENT, WorkTargetReferenceType.ORCHID_GROUP, true),
	REPOT(null, WorkTypeWorkflow.STRUCTURE_CHANGE, WorkTargetReferenceType.ORCHID_GROUP, true),
	DIVIDE("DIVIDE", WorkTypeWorkflow.STRUCTURE_CHANGE, WorkTargetReferenceType.ORCHID_GROUP, true),
	MERGE("MERGE", WorkTypeWorkflow.STRUCTURE_CHANGE, WorkTargetReferenceType.ORCHID_GROUP, true),
	DISCARD("DISCARD", WorkTypeWorkflow.DISCARD, WorkTargetReferenceType.ORCHID_GROUP, true),
	MULTI_CREATE(null, WorkTypeWorkflow.GENERIC, WorkTargetReferenceType.ORCHID_GROUP, true),
	CORRECTION(null, WorkTypeWorkflow.GENERIC, WorkTargetReferenceType.ORCHID_GROUP, true);

	private static final Map<String, WorkTypeDefinition> BY_CODE = Arrays.stream(values())
			.collect(Collectors.toUnmodifiableMap(Enum::name, Function.identity()));

	private final String handlerOverride;
	private final WorkTypeWorkflow workflow;
	private final WorkTargetReferenceType targetSource;
	private final boolean manualRegistrationAllowed;

	public static WorkTypeDefinition forCode(String code) {
		return code == null ? GENERIC : BY_CODE.getOrDefault(code, GENERIC);
	}

	public String handlerCode(WorkTypeTemplate template) {
		// Existing code overrides take precedence; other codes retain their stored template's handler.
		return handlerOverride == null ? template.handlerCode() : handlerOverride;
	}

	public WorkTypeWorkflow workflow() {
		return workflow;
	}

	public WorkTargetReferenceType targetSource() {
		return targetSource;
	}

	public boolean allowsManualRegistration() {
		return manualRegistrationAllowed;
	}

	public boolean hasDedicatedRegistration() {
		return workflow != WorkTypeWorkflow.GENERIC;
	}

	public boolean supportsPeriodPlanning() {
		return hasDedicatedRegistration() && workflow != WorkTypeWorkflow.POTTING;
	}

	public boolean supportsStructureExecution() {
		return workflow == WorkTypeWorkflow.STRUCTURE_CHANGE || workflow == WorkTypeWorkflow.MOVEMENT;
	}

	public static Set<String> requiredHandlerCodes() {
		return Stream.concat(
				Arrays.stream(values()).map(definition -> definition.handlerOverride).filter(Objects::nonNull),
				Arrays.stream(WorkTypeTemplate.values()).map(WorkTypeTemplate::handlerCode))
				.collect(Collectors.toUnmodifiableSet());
	}
}
